File: data/sysinfo.js
fetch("/sysinfo")
.then(r => r.json())
.then(data => {
    const tbody = document.querySelector("#sysinfo tbody");

    tbody.innerHTML = `
        <tr><th colspan="2">System</th></tr>
        <tr><td>ID</td><td>${data.id}</td></tr>
        <tr><td>Version</td><td>${data.version}</td></tr>
        <tr><td>Date/time</td><td>${formatDate(data.datetime)}</td></tr>
        <tr><td>Uptime</td><td>${formatDuration(data.uptime)}</td></tr>

        <tr><th colspan="2">CPU</th></tr>
        <tr><td>Chip</td><td>${data.cpu.chipmodel}</td></tr>
        <tr><td>Frequency</td><td>${data.cpu.frequency} MHz</td></tr>
        <tr><td>Temperature</td><td>${data.cpu.temperature} deg C</td></tr>

        <tr><th colspan="2">Memory</th></tr>
        <tr><td>Total heap</td><td>${data.heap.total}</td></tr>
        <tr><td>Free heap</td><td>${data.heap.free}</td></tr>

        <tr><th colspan="2">Wi-Fi</th></tr>
        <tr><td>SSID</td><td>${data.wifi.ssid}</td></tr>
        <tr><td>BSSID</td><td>${data.wifi.bssid}</td></tr>
        <tr><td>Channel</td><td>${data.wifi.channel}</td></tr>
        <tr><td>Signal</td><td>${data.wifi.rssi} dBm</td></tr>
    `;
});

File: sysinfo.cpp
#include <string.h>             // for strcpy

#include <Arduino.h>
#include <ArduinoJson.h>
#include "LittleFS.h"
#include <WiFi.h>

#include "version.h"
#include "sysinfo.h"

static const char *node_id;

static void handleGetRequest(AsyncWebServerRequest *request)
{
    AsyncResponseStream *response = request->beginResponseStream("application/json");

    StaticJsonDocument < 1024 > doc;
    doc["id"] = node_id;
    doc["version"] = GIT_VERSION;
    doc["datetime"] = time(nullptr);
    doc["uptime"] = millis() / 1000;

    JsonObject cpu = doc["cpu"].to < JsonObject > ();
    cpu["chipmodel"] = ESP.getChipModel();
    char temp[10];
    snprintf(temp, sizeof(temp), "%.1f", temperatureRead());
    cpu["frequency"] = ESP.getCpuFreqMHz();
    cpu["temperature"] = temp;

    JsonObject heap = doc["heap"].to < JsonObject > ();
    heap["total"] = ESP.getHeapSize();
    heap["free"] = ESP.getFreeHeap();

    JsonObject wifi = doc["wifi"].to < JsonObject > ();
    wifi["ssid"] = WiFi.SSID();
    wifi["bssid"] = WiFi.BSSIDstr();
    wifi["channel"] = WiFi.channel();
    wifi["rssi"] = WiFi.RSSI();

    serializeJson(doc, *response);
    request->send(response);
}

void sysinfo_begin(const char *id)
{
    node_id = id;
}

void sysinfo_serve(AsyncWebServer & server, const char *path)
{
    // register ourselves with the server
    server.on(path, HTTP_GET, handleGetRequest);
}
